util/clusterauth/clusterauth.go

```diff
@@ -215,39 +217,111 @@ func InstallClusterManagerRBAC(clientset kubernetes.Interface, ns string, namesp
-// kubernetes.io/service-account-token, and return it's token if found.
+// GetServiceAccountBearerToken determines if a ServiceAccount has a
+// bearer token secret to use or if a secret should be created. It then
+// waits for the secret to have a bearer token if a secret needs to
+// be created and returns the token in encoded base64.
 func GetServiceAccountBearerToken(clientset kubernetes.Interface, ns string, sa string) (string, error) {
-	var serviceAccount *corev1.ServiceAccount
+	secretName, err := createServiceAccountTokenSecret(clientset, sa, ns)
+	if err != nil {
+		return "", err
+	}
+
 	var secret *corev1.Secret
-	var err error
 	err = wait.Poll(500*time.Millisecond, 30*time.Second, func() (bool, error) {
-		serviceAccount, err = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+		secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), secretName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("failed to retrieve secret %q: %v", secretName, err)
+		}
+
+		_, ok := secret.Data["token"]
+		if !ok {
+			return false, fmt.Errorf("secret %q for serviceaccount %q did not have a token", secret.Name, sa)
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to get token for serviceaccount %q: %w", sa, err)
+	}
+
+	return string(secret.Data["token"]), nil
+}
+
+// createServiceAccountTokenSecret will check if a ServiceAccount
+// already has a kubernetes.io/service-account-token secret associated
+// with it or creates one if the ServiceAccount doesn't have one. This
+// was added to help add k8s v1.24+ clusters.
+func createServiceAccountTokenSecret(clientset kubernetes.Interface, sa, ns string) (string, error) {
+	// Wait for sa to have secret, but don't wait too
+	// long for 1.24+ clusters
+	var serviceAccount *corev1.ServiceAccount
+	err := wait.Poll(500*time.Millisecond, 5*time.Second, func() (bool, error) {
+		var getErr error
+		serviceAccount, getErr = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+		if getErr != nil {
+			return false, fmt.Errorf("failed to get serviceaccount %q: %w", sa, getErr)
+		}
+		if len(serviceAccount.Secrets) == 0 {
+			return false, nil
+		}
+		return true, nil
+	})
+	if serviceAccount == nil {
+		if err != nil {
+			return "", fmt.Errorf("failed to create serviceaccount token secret: %w", err)
 		}
-		// Scan all secrets looking for one of the correct type:
-		for _, oRef := range serviceAccount.Secrets {
-			var getErr error
-			secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), oRef.Name, metav1.GetOptions{})
+		log.Errorf("Unexpected nil serviceaccount '%s/%s' with no error returned", ns, sa)
+		return "", fmt.Errorf("failed to create serviceaccount token secret: nil serviceaccount returned for '%s/%s' with no error", ns, sa)
+	}
+
+	if len(serviceAccount.Secrets) != 0 {
```

Comment:
I think the `len` check is unnecessary - if `Secrets` has no items, the loop will just be a no-op.